Stats rewrite

rr- opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

Brings over the following features to TR2X:
- death counter
- pickup counts
- max pickup counts
- max kill counts
- max secret counts (not visible in the OG as we display dragons instead)

Resolves #2578.

Internally, this is a full rewrite of the stats module and thus it's a pretty heavyweight change.

The maximum counts are calculated right after the injections load, once per level, using real data. That information goes straight into `RESUME_INFO` for the corresponding level. I was careful to make sure resets and carryovers of `RESUME_INFO` don't wipe out the maximum count data (see: `M_CopyResumeInfo`). I think the whole setup turned out quite elegant, as all the scanning happens neatly in one place now, from handling the dragon bits to walking the floor data and everything in between.
Because the maximum counts are all computed unconditionally right after the level finishes loading, we no longer load/store the maximum counts from/to either save file formats.

I also watered down the description of the statistics dialog to keep it version-agnostic, skipping over the specific options. It's not perfect, but this wording lets us dodge yet another version branch.

Here’s a table comparing the counts the game reports for TR2 with the numbers from Stella’s walkthroughs. I’m not totally sure yet where the discrepancies come from and which version is right.

| #  | Title                    | File           |      | Pickups               |      | Kills                 |     | Secrets |
| -- | ------------------------ | -------------- | ---  | --------              | ---  | -----                 | --  | ------- |
| 0  | Lara's Home              | `assault.tr2`  | ✅1  | 1                     | ✅-  | -                     | ✅- | -       |
| 1  | The Great Wall           | `wall.tr2`     | ✅14 | 14                    | ✅23 | 23                    | ✅3 | 3       |
| 2  | Venice                   | `boat.tr2`     | ✅30 | 30                    | ✅24 | 24                    | ✅3 | 3       |
| 3  | Bartoli's Hideout        | `venice.tr2`   | ✅28 | 28                    | ✅37 | 37                    | ✅3 | 3       |
| 4  | Opera House              | `opera.tr2`    | ✅37 | 37                    | ❌47 | 46                    | ✅3 | 3       |
| 5  | Offshore Rig             | `rig.tr2`      | ✅31 | 31                    | ✅20 | 20                    | ✅3 | 3       |
| 6  | Diving Area              | `platform.tr2` | ✅39 | 39                    | ✅34 | 34                    | ✅3 | 3       |
| 7  | 40 Fathoms               | `unwater.tr2`  | ✅14 | 14                    | ✅16 | 16                    | ✅3 | 3       |
| 8  | Wreck of the Maria Doria | `keel.tr2`     | ✅41 | 41                    | ❌36 | 35                    | ✅3 | 3       |
| 9  | Living Quarters          | `living.tr2`   | ✅16 | 16                    | ✅21 | 21                    | ✅3 | 3       |
| 10 | The Deck                 | `deck.tr2`     | ✅35 | 35                    | ✅30 | 30                    | ✅3 | 3       |
| 11 | Tibetan Foothills        | `skidoo.tr2`   | ✅31 | 31                    | ✅33 | 33                    | ✅3 | 3       |
| 12 | Barkhang Monastery       | `monastry.tr2` | ❌46 | 49                    | ❌31 | 30 / 52<sup>[1]</sup> | ✅3 | 3       |
| 13 | Catacombs of the Talion  | `catacomb.tr2` | ❌40 | 39                    | ✅33 | 33                    | ✅3 | 3       |
| 14 | Ice Palace               | `icecave.tr2`  | ✅33 | 33                    | ✅22 | 22                    | ✅3 | 3       |
| 15 | Temple of Xian           | `emprtomb.tr2` | ✅40 | 39 / 40<sup>[2]</sup> | ✅37 | 37                    | ✅3 | 3       |
| 16 | Floating Islands         | `floating.tr2` | ✅39 | 39                    | ✅26 | 26                    | ✅3 | 3       |
| 17 | The Dragon's Lair        | `xian.tr2`     | ✅25 | 25                    | ✅11 | 11                    | ✅- | -       |
| 18 | Home Sweet Home          | `house.tr2`    | ✅45 | 45                    | ✅16 | 16                    | ✅- | -       |

I'd appreciate testing of the following areas:

- the validity of the numbers above,
- order of the dragons,
- support for more than 3 dragons,
- saving and reloading,
- fiddling with the stats dialogs modes,
- fiddling with the death counter on/off,

This change affects both TR1 and TR2 equally.
Related test levels:

- `test-secrets-a.tr2`
- `test-secrets-b.tr2`
- `test-secrets-c.tr2`
- `test-secrets-d.tr2`
- `test-secrets-e.tr2`
- `test-secrets-f.tr2`
- `test-secrets-flipmaps.tr2`
- `test-secrets-g.tr2`

After this PR, I'd like to focus on merging the UI parts in both games.

rr- committed:
savegame: drop support for tr1x 2.6 and 2.7 saves

rr- committed:
stats: rewrite

rr- committed:
tools/inspect-save: read extended header

aredfan commented:
**Issues:**

- [x] TR2: In the gym, the stopwatch has the level stats instead of best times.  This issue appears to be present in develop.
- [x] TR2: The pickups stats remain on `0` when Lara picks up something. e.g. the medi pack in TGW.
- [ ] ~~TR2: In Statistics details, only the `Full` option is effective.~~

rr- commented:
I fixed the Assault Coure stats gym in https://github.com/LostArtefacts/TRX/pull/4183.
I fixed the pickup bug.
As for the detail modes, they differences are subtle, but I can see them on my end:

<img width="2664" height="1473" alt="20251028_092354_The_Great_Wall" src="https://github.com/user-attachments/assets/0ef3bdcb-bf3e-4690-9d4b-29150e129a91" />
<img width="2664" height="1473" alt="20251028_092359_The_Great_Wall" src="https://github.com/user-attachments/assets/5f62db20-a9db-45f4-b32a-473984cbd27e" />
<img width="2664" height="1473" alt="20251028_092403_The_Great_Wall" src="https://github.com/user-attachments/assets/6705a428-726f-4ee6-be54-5d741b08eff3" />

rr- commented:
I fixed one bug where the secret rewards wouldn't contribute towards collected pickup count, and with the help of Lahm I have found and adjusted a few untriggerable/unreachable enemies. Current standing – only pickups are wonky:

## Tomb Raider I

| #  | Title                  | File           |      | Pickups               |      | Kills                 |     | Secrets         | Save Crystals |
| -- | ---------------------- | -------------- | ---- | --------------------- | ---- | --------------------- | --- | --------------- | ------------- |
| 0  | Lara's Home            | `gym.phd`      | ✅-  | -                     | ✅-  | -                     | ✅- | -               | -             |
| 1  | Caves                  | `level1.phd`   | ✅7  | 7                     | ✅14 | 14                    | ✅3 | 3               | 1             |
| 2  | City of Vilcabamba     | `level2.phd`   | ✅13 | 13                    | ✅29 | 29                    | ✅3 | 3               | 2             |
| 3  | Lost Valley            | `level3a.phd`  | ✅16 | 16                    | ✅13 | 13                    | ✅5 | 5               | 3             |
| 4  | Tomb of Qualopec       | `level3b.phd`  | ✅8  | 8                     | ✅8  | 7 / 8<sup>[1]</sup>   | ✅3 | 3               | 3             |
| 5  | St. Francis' Folly     | `level4.phd`   | ✅19 | 19                    | ✅23 | 23                    | ✅4 | 4               | 5             |
| 6  | Colosseum              | `level5.phd`   | ✅14 | 14                    | ✅27 | 26 / 27<sup>[2]</sup> | ✅3 | 3               | 4             |
| 7  | Palace Midas           | `level6.phd`   | ✅22 | 22 / 23<sup>[3]</sup> | ✅43 | 43                    | ✅3 | 3               | 8             |
| 8  | The Cistern            | `level7a.phd`  | ✅28 | 28                    | ✅34 | 34                    | ✅3 | 3               | 5             |
| 9  | Tomb of Tihocan        | `level7b.phd`  | ✅26 | 26                    | ✅17 | 17                    | ✅2 | 2               | 5             |
| 10 | City of Khamoon        | `level8a.phd`  | ✅24 | 24                    | ✅15 | 14 / 15<sup>[4]</sup> | ✅3 | 3               | 4             |
| 11 | Obelisk of Khamoon     | `level8b.phd`  | ✅38 | 38                    | ✅16 | 16                    | ✅3 | 3               | 7             |
| 12 | Sanctuary of the Scion | `level8c.phd`  | ✅29 | 29                    | ✅15 | 15                    | ✅1 | 1               | 7             |
| 13 | Natla's Mines          | `level10a.phd` | ✅30 | 30                    | ✅3  | 3                     | ✅3 | 3               | 7             |
| 14 | Atlantis               | `level10b.phd` | ✅50 | 50                    | ✅32 | 32                    | ✅3 | 3               | 7             |
| 15 | The Great Pyramid      | `level10c.phd` | ✅31 | 31                    | ✅6  | 6                     | ✅3 | 3<sup>[5]</sup> | 4             |

- <sup>[1]</sup>: extra mummy kill in the Tomb of Qualopec.
- <sup>[2]</sup>: bat with a missing trigger in the OG.
- <sup>[3]</sup>: unobtainable medipack in the Midas hub.
- <sup>[4]</sup>: additional mummy in the PS1 version.
- <sup>[5]</sup>: broken secret trigger in some versions of the game.

## Tomb Raider I: Unfinished Business

| #  | Title                | File        |      | Pickups |      | Kills |      | Secrets |
| -- | -------------------- | ----------- | ---- | ------- | ---- | ----- | ---- | ------- |
| 1  | Return to Egypt      | `egypt.phd` | ✅53 | 53      | ✅41 | 41    | ✅3  | 3       |
| 2  | Temple of the Cat    | `cat.phd`   | ✅63 | 63      | ✅44 | 44    | ✅4  | 4       |
| 3  | Atlantean Stronghold | `end.phd`   | ✅63 | 63      | ✅31 | 31    | ✅2  | 2       |
| 4  | The Hive             | `end2.phd`  | ✅60 | 60      | ✅41 | 41    | ✅1  | 1       |

## Tomb Raider II

| #  | Title                    | File           |      | Pickups               |      | Kills                 |     | Secrets |
| -- | ------------------------ | -------------- | ---- | --------------------- | ---- | --------------------- | --- | ------- |
| 0  | Lara's Home              | `assault.tr2`  | ✅1  | 1                     | ✅-  | -                     | ✅- | -       |
| 1  | The Great Wall           | `wall.tr2`     | ✅14 | 14                    | ✅23 | 23                    | ✅3 | 3       |
| 2  | Venice                   | `boat.tr2`     | ✅30 | 30                    | ✅24 | 24                    | ✅3 | 3       |
| 3  | Bartoli's Hideout        | `venice.tr2`   | ✅28 | 28                    | ✅37 | 37                    | ✅3 | 3       |
| 4  | Opera House              | `opera.tr2`    | ✅37 | 37                    | ✅46 | 46                    | ✅3 | 3       |
| 5  | Offshore Rig             | `rig.tr2`      | ✅31 | 31                    | ✅20 | 20                    | ✅3 | 3       |
| 6  | Diving Area              | `platform.tr2` | ✅39 | 39                    | ✅34 | 34                    | ✅3 | 3       |
| 7  | 40 Fathoms               | `unwater.tr2`  | ✅14 | 14                    | ✅16 | 16                    | ✅3 | 3       |
| 8  | Wreck of the Maria Doria | `keel.tr2`     | ✅41 | 41                    | ✅35 | 35 / 36<sup>[1]</sup> | ✅3 | 3       |
| 9  | Living Quarters          | `living.tr2`   | ✅16 | 16                    | ✅21 | 21                    | ✅3 | 3       |
| 10 | The Deck                 | `deck.tr2`     | ✅35 | 35                    | ✅30 | 30                    | ✅3 | 3       |
| 11 | Tibetan Foothills        | `skidoo.tr2`   | ✅31 | 31                    | ✅33 | 33                    | ✅3 | 3       |
| 12 | Barkhang Monastery       | `monastry.tr2` | ❌46 | 49                    | ✅30 | 30 / 52<sup>[2]</sup> | ✅3 | 3       |
| 13 | Catacombs of the Talion  | `catacomb.tr2` | ✅39 | 39                    | ✅33 | 33                    | ✅3 | 3       |
| 14 | Ice Palace               | `icecave.tr2`  | ✅33 | 33                    | ✅22 | 22                    | ✅3 | 3       |
| 15 | Temple of Xian           | `emprtomb.tr2` | ✅39 | 39 / 40<sup>[3]</sup> | ✅37 | 37                    | ✅3 | 3       |
| 16 | Floating Islands         | `floating.tr2` | ✅39 | 39                    | ✅26 | 26                    | ✅3 | 3       |
| 17 | The Dragon's Lair        | `xian.tr2`     | ✅25 | 25                    | ✅11 | 11                    | ✅- | -       |
| 18 | Home Sweet Home          | `house.tr2`    | ✅45 | 45                    | ✅16 | 16                    | ✅- | -       |

- <sup>[1]</sup>: shark in an unreachable room (possible to kill with extreme patience)
- <sup>[2]</sup>: hostiles vs. hostiles and allies
- <sup>[3]</sup>: unobtainable medipack above the waterfall lake

## Tomb Raider II: The Golden Mask

| #  | Title               | File         |      | Pickups               |      | Kills                 |     | Secrets   |
| -- | ------------------- | ------------ | ---- | --------------------- | ---- | --------------------- | --- | --------- |
| 1  | The Cold War        | `level1.tr2` | ✅71 | 71                    | ✅39 | 39 / 44<sup>[1]</sup> | ✅3 | 3         |
| 2  | Fool's Gold         | `level2.tr2` | ✅69 | 69                    | ✅62 | 62                    | ✅3 | 3         |
| 3  | Furnace of the Gods | `level3.tr2` | ✅64 | 64                    | ✅38 | 38 / 41<sup>[1]</sup> | ✅3 | 3         |
| 4  | Kingdom             | `level4.tr2` | ✅50 | 50 / 52<sup>[2]</sup> | ✅25 | 25 / 29<sup>[1]</sup> | ✅3 | 3         |
| 5  | Nightmare in Vegas  | `level5.tr2` | ✅75 | 75                    | ✅23 | 23                    | ✅3 | 3         |

- <sup>[1]</sup>: hostiles vs. hostiles and allies
- <sup>[2]</sup>: additional drops from allies

aredfan commented:
Thank you for explaining, @rr-, my eyes have failed me this morning. 😬 

- [x] TR2: Ammo Used issue:

> <details><summary>The detailed stats has `Ammo Used` which is correct.</summary>
> 
> ![20251028_102948_Floating_Islands](https://github.com/user-attachments/assets/8b2bd7ab-31e2-4a82-b237-f9d08fb5a058)
> 
> </details>
> 
> 
> <details><summary>The full stats has `Ammo Used` which is incorrect. TR1X displays this as `Ammo Hits/Used` instead.</summary>
> 
> ![20251028_102956_Floating_Islands](https://github.com/user-attachments/assets/82a75297-eedd-4cf1-af60-d8eff85d3789)
> 
> </details>
> 
> <details><summary>Full stats in TR1X, for reference</summary>
> 
> ![20251028_083123_Caves](https://github.com/user-attachments/assets/4e895955-711b-4ad2-bca4-3f81c6e32a86)
> 
> </details>

- [x] TR2: the pickup stats doesn't seem to get saved.

> 1. `/play 1`
> 2. `/tp medi`
> 3. `/save 1`
> 4. `/load 1`
> 5. open the stopwatch to view the stats.

aredfan commented:
- [x] TR2: It looks like the pickup stats doesn't reset between levels.  To reproduce, pick up an item in TGW, then use level skip to reach Venice and open the stopwatch.

rr- commented:
> * [ ]  TR2: It looks like the pickup stats doesn't reset between levels.  To reproduce, pick up an item in TGW, then use level skip to reach Venice and open the stopwatch.

Fixed :relieved:

rr- commented:
(Just squashed all commits together for tidiness.)

rr- committed:
misc: fix Barkhang pickup count

lahm86 reviewed (approved):
Great!
